docs/asciidoc/static/include/pluginbody.asciidoc

```diff
@@ -0,0 +1,970 @@
+codec.
+
+There must be a `yield` statement as part of the `decode` method which will
+return decoded events to the pipeline.
+endif::decode_method[]
+
+// /////////////////////////////////////////////////////////////////////////////
+// If encode_method is defined (should only be for codec plugin page)
+// /////////////////////////////////////////////////////////////////////////////
+ifdef::encode_method[]
+==== `encode` Method
+[source,ruby]
+[subs="attributes"]
+----------------------------------
+  public
+  def encode(event)
+    @on_event.call(event, event["message"].to_s + @append + NL)
+  end # def encode
+----------------------------------
+The `encode` method takes an event and serializes it (_encodes_) into another
+format.  Good examples of `encode` methods include the simple https://github.com/logstash-plugins/logstash-codec-plain/blob/master/lib/logstash/codecs/plain.rb#L39-L46[plain]
+codec, the slightly more involved https://github.com/logstash-plugins/logstash-codec-msgpack/blob/master/lib/logstash/codecs/msgpack.rb#L38-L46[msgpack]
+codec, and even an https://github.com/logstash-plugins/logstash-codec-avro/blob/master/lib/logstash/codecs/avro.rb#L38-L45[avro]
+codec.
+
+In most cases, your `encode` method should have an `@on_event.call()` statement.
+This call will output data per event in the described way.
+endif::encode_method[]
+
+// /////////////////////////////////////////////////////////////////////////////
+// If run_method is defined (should only be for input plugin page)
+// /////////////////////////////////////////////////////////////////////////////
+ifdef::run_method[]
+==== `run` Method
+The `pass:attributes[{pluginname}]` input plugin has the following `run` Method:
+[source,ruby]
+[subs="attributes"]
+----------------------------------
+  def run(queue)
+    Stud.interval(@interval) do
+      event = LogStash::Event.new("message" => @message, "host" => @host)
+      decorate(event)
+      queue << event
+    end # loop
+  end # def run
+----------------------------------
+The `run` method is where a stream of data from an input becomes an event.
+
+The stream can be plain or generated as with the
+https://github.com/logstash-plugins/logstash-input-heartbeat/blob/master/lib/logstash/inputs/heartbeat.rb#L43-L61[heartbeat]
+input plugin.  In these cases, though no codec is used,
+https://github.com/logstash-plugins/logstash-input-heartbeat/blob/master/lib/logstash/inputs/heartbeat.rb#L17[a default codec]
+must be set in the code to avoid errors.
+
+Here's another example `run` method:
+[source,ruby]
+[subs="attributes"]
+----------------------------------
+  def run(queue)
+    while true
+      begin
+        # Based on some testing, there is no way to interrupt an IO.sysread nor
+        # IO.select call in JRuby.
+        data = $stdin.sysread(16384)
+        @codec.decode(data) do |event|
+          decorate(event)
```

Comment:
That's not a bad idea, but I will not do so unless the code in the stdin plugin—which this example was drawn from—is also changed so they match.